### DISCUSSION TITLE
How to add Accept version header in browesable API?

### DISCUSSION THREAD

**Discussion by kisamoto** [encode/django-rest-framework#8449] [Question & Answer]
Hi all,

Trying to build an API using best practices and noticed the recommendation to use an `Accept` [header](https://www.django-rest-framework.org/api-guide/versioning/#acceptheaderversioning) to version my API. 

However when I try to access my API with the builtin browsable API, I get an invalid version error:

```json
{
    "detail": "Invalid version in \"Accept\" header."
}
```

How would I extend the browesable API to include a version header (ideally a user-selectable dropdown version but a hard-coded default latest version would work for now)?

Or what would be the recommended practice here?

Thanks

**Comment by hashlash** [comment]
I don't think DRF has a feature for it, and _browsers don't generally let you set custom headers_ (quoted from [an SO answer](https://stackoverflow.com/a/24403244/7168030)).

Searching on StackOverflow ([1](https://stackoverflow.com/a/44014014/7168030), [2](https://stackoverflow.com/a/27012197/7168030), [3](https://stackoverflow.com/a/64594311/7168030)), it seems using [modheader](https://chrome.google.com/webstore/detail/modheader/idgpnmonknjnojddfkpgkljpfnnfcklj?hl=en) browser extension is a somewhat popular way to add your custom header.

However, you could also override the browsable API script to include a custom header yourself. I believe the script lies on https://github.com/encode/django-rest-framework/blob/master/rest_framework/static/rest_framework/js/ajax-form.js.

**Comment by voxvici** [comment]
Check this https://www.django-rest-framework.org/api-guide/versioning/, usually api version is included in URL param

**Comment by kisamoto** [comment]
I am aware of the docs but from them it says:

"Versioning based on accept headers is generally considered
<http://blog.steveklabnik.com/posts/2011-07-03-nobody-understands-rest-or-http#i_want_my_api_to_be_versioned>
as best practice
<https://github.com/interagent/http-api-design/blob/master/en/foundations/require-versioning-in-the-accepts-header.md>
"

As I want to follow best practices I was hoping the rest framework
browsable API would support this out of the box but I think I will need to
extend it.

I do not wish to have clients using the test harness to install a new
extension to add the header.

**Reply by voxvici** [reply]
It does support it out of the box..read the docs

**Comment by kisamoto** [comment]
Thank you - I will look into extending the browsable API and maybe open a
feature request to have it implemented in the core.

Would be good to have the browsable API as close to feature compatible with
what we can build.

**Reply by voxvici** [reply]
There is every possible way of adding API version within django rest framework and for your header solution as well https://www.django-rest-framework.org/api-guide/versioning/#acceptheaderversioning

**Comment by kisamoto** [comment]
Yes - I have that implemented thanks.

The question is not how to add versioning to my API but how to get the
built-in browsable API to support this without adding an additional browser
extension.
